File: joss/comparisons/cat-1-column-schema-reference/07-pipeline-composition/example-pipeline-composition.py
"""
Error Class 7: Pipeline Composition Errors

In Python, multi-step pandas pipelines have no compile-time schema tracking.
Column names are strings — typos, references to renamed columns, and
post-aggregation column access all fail only at runtime.
"""
import pandas as pd

patients = pd.read_csv("fixtures/patients.csv")
encounters = pd.read_csv("fixtures/encounters.csv")
labs = pd.read_csv("fixtures/lab_results.csv")
meds = pd.read_csv("fixtures/medications.csv")

# ── ERROR 7a: Wrong column after rename ─────────────────────────────────
# RUNTIME: KeyError — 'department' was renamed to 'dept'
pipeline = encounters.rename(columns={"department": "dept"})
try:
    pipeline = pipeline[pipeline["department"] == "ICU"]
except KeyError:
    print("KeyError: 'department' was renamed — runtime only")

# ── ERROR 7b: Wrong column after multi-step transformation ──────────────
# RUNTIME: KeyError — after groupby/agg, original columns are gone
pipeline = (
    encounters
    .merge(labs, on=["encounter_id", "patient_id"])
    [["patient_id", "department", "test_name", "result_value"]]
    .groupby("patient_id")
    .agg(max_lab=("result_value", "max"))
    .reset_index()
)
try:
    pipeline["dept"] = pipeline["department"]  # KeyError
except KeyError:
    print("KeyError: 'department' gone after groupby — runtime only")

# ── CORRECT (but unchecked) pipeline ────────────────────────────────────
# This works, but no tool validates column names until execution.
# A typo anywhere in this chain fails silently or with a runtime error.
report = (
    encounters
    .merge(labs, on=["encounter_id", "patient_id"])
    .query("abnormal_flag == 'H'")
    .groupby("patient_id")
    .agg(
        abnormal_count=("lab_id", "count"),
        max_result=("result_value", "max"),
        tests=("test_name", lambda x: ", ".join(x.unique()))
    )
    .reset_index()
    .merge(patients, on="patient_id")
    .assign(full_name=lambda df: df["last_name"] + ", " + df["first_name"])
    [["full_name", "insurance_type", "abnormal_count", "max_result", "tests"]]
    .sort_values("abnormal_count", ascending=False)
)
print(report)

File: joss/comparisons/cat-1-column-schema-reference/07-pipeline-composition/example-pipeline-composition.R
# Error Class 7: Pipeline Composition Errors
#
# In R, multi-step tidyverse pipelines have no compile-time schema tracking.
# Column references use non-standard evaluation — typos and references to
# columns removed by earlier steps fail only at runtime.

library(tidyverse)

patients <- read_csv("fixtures/patients.csv")
encounters <- read_csv("fixtures/encounters.csv")
labs <- read_csv("fixtures/lab_results.csv")
meds <- read_csv("fixtures/medications.csv")

# ── ERROR 7a: Wrong column after rename ─────────────────────────────────
# RUNTIME ERROR: Column `department` doesn't exist — was renamed to `dept`.
encounters %>%
  rename(dept = department) %>%
  filter(department == "ICU")

# ── ERROR 7b: Wrong column after multi-step transformation ──────────────
# RUNTIME ERROR: Column `department` doesn't exist after summarise.
encounters %>%
  left_join(labs, by = c("encounter_id", "patient_id")) %>%
  select(patient_id, department, test_name, result_value) %>%
  group_by(patient_id) %>%
  summarise(max_lab = max(result_value)) %>%
  mutate(dept = department)  # Error — department gone after summarise

# ── CORRECT (but unchecked) pipeline ────────────────────────────────────
# This works, but R validates nothing until the pipeline executes.
# A typo anywhere in this chain fails only at runtime.
report <- encounters %>%
  left_join(labs, by = c("encounter_id", "patient_id")) %>%
  filter(abnormal_flag == "H") %>%
  group_by(patient_id) %>%
  summarise(
    abnormal_count = n(),
    max_result = max(result_value),
    tests = paste(unique(test_name), collapse = ", ")
  ) %>%
  left_join(patients, by = "patient_id") %>%
  mutate(full_name = paste0(last_name, ", ", first_name)) %>%
  select(full_name, insurance_type, abnormal_count, max_result, tests) %>%
  arrange(desc(abnormal_count))

print(report)
